dds/DCPS/XTypes/DynamicDataBase.cpp

```diff
@@ -296,32 +299,83 @@ DDS::MemberId DynamicDataBase::get_union_default_member(DDS::DynamicType* type)
   return default_branch;
 }
 
-bool DynamicDataBase::discriminator_selects_no_member(DDS::DynamicType* type, ACE_CDR::Long disc)
+DDS::ReturnCode_t DynamicDataBase::get_selected_union_branch(DDS::Int32 disc,
+  bool& found_selected_member, DDS::MemberDescriptor_var& selected_md) const
 {
-  const ACE_CDR::ULong members = type->get_member_count();
-  for (ACE_CDR::ULong i = 0; i < members; ++i) {
-    DDS::DynamicTypeMember_var member;
-    if (type->get_member_by_index(member, i) != DDS::RETCODE_OK) {
-      return false;
+  found_selected_member = false;
+  bool has_default = false;
+  DDS::ReturnCode_t rc = DDS::RETCODE_OK;
+  DDS::MemberDescriptor_var default_md;
+  for (DDS::UInt32 i = 0; i < type_->get_member_count(); ++i) {
+    DDS::DynamicTypeMember_var dtm;
+    rc = type_->get_member_by_index(dtm, i);
+    if (rc != DDS::RETCODE_OK) {
+      return rc;
     }
-    if (member->get_id() == DISCRIMINATOR_ID) {
+    if (dtm->get_id() == DISCRIMINATOR_ID) {
       continue;
     }
-    DDS::MemberDescriptor_var mdesc;
-    if (member->get_descriptor(mdesc) != DDS::RETCODE_OK) {
-      return false;
-    }
-    if (mdesc->is_default_label()) {
-      return false;
+    DDS::MemberDescriptor_var md;
+    rc = dtm->get_descriptor(md);
+    if (rc != DDS::RETCODE_OK) {
+      return rc;
     }
-    const DDS::UnionCaseLabelSeq& lseq = mdesc->label();
-    for (ACE_CDR::ULong lbl = 0; lbl < lseq.length(); ++lbl) {
-      if (lseq[lbl] == disc) {
-        return false;
+    bool found_matched_label = false;
+    const DDS::UnionCaseLabelSeq labels = md->label();
+    for (DDS::UInt32 j = 0; !found_matched_label && j < labels.length(); ++j) {
+      if (disc == labels[j]) {
+        found_matched_label = true;
       }
     }
+    if (found_matched_label) {
+      selected_md = md;
+      found_selected_member = true;
+      break;
+    }
+    if (md->is_default_label()) {
+      default_md = md;
+      has_default = true;
+    }
   }
-  return true;
+  if (!found_selected_member && has_default) {
+    selected_md = default_md;
+    found_selected_member = true;
+  }
+  return rc;
+}
+
+DDS::ReturnCode_t DynamicDataBase::get_selected_union_branch(
+  bool& found_selected_member, DDS::MemberDescriptor_var& selected_md)
+{
+  // TODO: Support UInt64 and Int64
```

Comment:
Sorta. It uses `get_int64_value`, but has to cast it down because the TypeObject is limited to storing discriminator values as `long`s. I think there's a spec issue for this. I created the casting `get_int64_value`, `get_uint64_value`, this comment  expecting this to get worked out eventually.

Reply:
I see. I wasn't sure what this comment is referring to.